File: aat-android/src/main/java/ch/bailu/aat/services/cache/osm_features/ObjMapFeatures.java
package ch.bailu.aat.services.cache.osm_features;


import java.util.ArrayList;

import ch.bailu.aat.preferences.map.SolidOsmFeaturesList;
import ch.bailu.aat.util.filter_list.AbsFilterList;
import ch.bailu.aat.util.filter_list.KeyList;
import ch.bailu.aat.util.filter_list.ListEntry;
import ch.bailu.aat_lib.app.AppContext;
import ch.bailu.aat_lib.dispatcher.AppBroadcaster;
import ch.bailu.aat_lib.service.background.BackgroundTask;
import ch.bailu.aat_lib.service.cache.Obj;
import ch.bailu.aat_lib.service.cache.OnObject;

public final class ObjMapFeatures extends Obj {

    public static final String ID_FULL = ObjMapFeatures.class.getSimpleName();
    public static final String ID_SMALL = ObjMapFeatures.class.getSimpleName()+"/s";

    private boolean isLoaded = false;
    private long size = 0;

    private class List {
        private final ArrayList<MapFeaturesListEntry> list = new ArrayList<>(50);

        public synchronized void clear() {
            list.clear();
        }

        public synchronized void sync(AbsFilterList<ListEntry> f) {
            for (int i = f.sizeAll(); i < list.size(); i++) {
                f.add(list.get(i));
            }
        }

        public synchronized void add(MapFeaturesListEntry d) {
            list.add(d);
        }
    }

    private final List list = new List();


    public ObjMapFeatures(String ID) {
        super(ID);
    }


    @Override
    public void onInsert(AppContext sc) {
        super.onInsert(sc);

        sc.getServices().getBackgroundService().process(new ListLoader(getID()));
    }

    @Override
    public void onDownloaded(String id, String url, AppContext sc) {}

    @Override
    public void onChanged(String id, AppContext sc) {}

    @Override
    public long getSize() {
        return size;
    }


    public void syncList(AbsFilterList<ListEntry> filterList) {
        list.sync(filterList);
    }


    @Override
    public boolean isReadyAndLoaded() {
        return isLoaded;
    }


    public static class ListLoader extends BackgroundTask implements MapFeaturesParser.OnHaveFeature {
        private final String ID;

        private AppContext appContext;

        private ObjMapFeatures owner = null;
        public ListLoader(String id) {
            ID = id;
        }

        public KeyList keyList;

        private int doBroadcast = 0;

        @Override
        public long bgOnProcess(final AppContext appContext) {
            ListLoader.this.appContext = appContext;
            final long[] size = {0};
            new OnObject(appContext, ID, ObjMapFeatures.class) {
                @Override
                public void run(Obj handle) {
                    owner = (ObjMapFeatures) handle;

                    parseMapFeatures();

                    owner.isLoaded = true;
                    ListLoader.this.appContext.getBroadcaster().broadcast(
                            AppBroadcaster.FILE_CHANGED_INCACHE, ID);

                    size[0] = owner.size;
                    owner = null;
                    ListLoader.this.appContext = null;
                }
            };
            return size[0];
        }


        private void parseMapFeatures() {
            keyList = SolidOsmFeaturesList.getKeyList(ID);
            new MapFeaturesParser(appContext.getAssets(), this);
        }


        @Override
        public boolean onParseFile(String file) {
            return keyList.isEmpty() || keyList.hasKey(file.toLowerCase());

        }

        @Override
        public void onHaveFeature(MapFeaturesParser parser) {
            MapFeaturesListEntry d = new MapFeaturesListEntry(parser);

            owner.size += d.length() * 2;
            owner.list.add(d);

            doBroadcast++;
            if (doBroadcast > 10) {
                doBroadcast = 0;
                appContext.getBroadcaster().broadcast(AppBroadcaster.FILE_CHANGED_INCACHE, ID);
            }
        }
    }

    public static class Factory extends Obj.Factory {
        final String ID;

        public Factory(String id) {
            ID = id;
        }

        @Override
        public Obj factory(String id, AppContext cs) {
            return new ObjMapFeatures(ID);
        }
    }
}

File: aat-android/src/main/kotlin/ch/bailu/aat/services/cache/osm_features/ObjMapFeatures.kt
package ch.bailu.aat.services.cache.osm_features

import ch.bailu.aat.preferences.map.SolidOsmFeaturesList.Companion.getKeyList
import ch.bailu.aat_lib.app.AppContext
import ch.bailu.aat_lib.broadcaster.AppBroadcaster
import ch.bailu.aat_lib.lib.filter_list.AbsFilterList
import ch.bailu.aat_lib.lib.filter_list.AbsListItem
import ch.bailu.aat_lib.service.background.BackgroundTask
import ch.bailu.aat_lib.service.cache.Obj
import ch.bailu.aat_lib.service.cache.OnObject

class ObjMapFeatures(id: String) : Obj(id) {
    private var isLoaded = false
    private var size: Long = 0

    private inner class List {
        private val list = ArrayList<MapFeaturesListItem>(50)

        @Synchronized
        fun sync(f: AbsFilterList<AbsListItem>) {
            for (i in f.sizeAll() until list.size) {
                f.add(list[i])
            }
        }

        @Synchronized
        fun add(d: MapFeaturesListItem) {
            list.add(d)
        }
    }

    private val list: List = List()

    override fun onInsert(appContext: AppContext) {
        super.onInsert(appContext)
        appContext.services.getBackgroundService().process(ListLoader(getID()))
    }

    override fun onDownloaded(id: String, url: String, appContext: AppContext) {}
    override fun onChanged(id: String, appContext: AppContext) {}
    override fun getSize(): Long {
        return size
    }

    fun syncList(filterList: AbsFilterList<AbsListItem>) {
        list.sync(filterList)
    }

    override fun isReadyAndLoaded(): Boolean {
        return isLoaded
    }

    class ListLoader(private val id: String) : BackgroundTask() {
        private var doBroadcast = 0

        override fun bgOnProcess(appContext: AppContext): Long {
            var size = 0L

            object : OnObject(appContext, id, ObjMapFeatures::class.java) {
                override fun run(handle: Obj) {
                    if (handle is ObjMapFeatures) {
                        parseMapFeatures(appContext, handle)
                        handle.isLoaded = true
                        appContext.broadcaster.broadcast(AppBroadcaster.FILE_CHANGED_INCACHE, id)
                        size = handle.size
                    }
                }
            }
            return size
        }

        private fun parseMapFeatures(appContext: AppContext, handle: ObjMapFeatures) {
            val keyList = getKeyList(id)

            MapFeaturesParser(appContext.assets, { file ->
                    keyList.isEmpty || keyList.hasKey(file)
                },
                { parser ->
                    val d = MapFeaturesListItem(parser)
                    handle.size += d.length() * 2L
                    handle.list.add(d)
                    doBroadcast++
                    if (doBroadcast > 10) {
                        doBroadcast = 0
                        appContext.broadcaster.broadcast(AppBroadcaster.FILE_CHANGED_INCACHE, id)
                }
            })
        }
    }

    class Factory(val id: String) : Obj.Factory() {
        override fun factory(id: String, appContext: AppContext): Obj {
            return ObjMapFeatures(this.id)
        }
    }

    companion object {
        val ID_FULL: String = ObjMapFeatures::class.java.simpleName
        val ID_SMALL = ObjMapFeatures::class.java.simpleName + "/s"
    }
}
